lib/Dialect/Seq/Transforms/LowerSeqToSV.cpp

```diff
@@ -364,6 +337,125 @@ void FirRegLower::initialise(OpBuilder &regBuilder, OpBuilder &initBuilder,
   recurse(type, "");
 }
 
+bool FirRegLower::isInitialisePreset(sv::RegOp reg) {
+  return module->hasAttr("firrtl.random_init_width") &&
+         reg->hasAttr("firrtl.random_init_start") &&
+         reg->hasAttr("firrtl.random_init_end");
+}
+
+void FirRegLower::initialisePreset(OpBuilder &regBuilder,
+                                   OpBuilder &initBuilder, sv::RegOp reg) {
+  // Extract the required random initialisation width and compute the number of
+  // `RANDOM calls required to fill it up, as well as the final register width.
+  auto randomInitWidthAttr =
+      module->getAttrOfType<IntegerAttr>("firrtl.random_init_width");
+  assert(randomInitWidthAttr &&
+         "firrtl.random_init_width required for preset initialisation");
+
+  uint64_t randomInitWidth = randomInitWidthAttr.getUInt();
+  uint64_t numRandomSources = llvm::divideCeil(randomInitWidth, randomWidth);
+  uint64_t randomRegWidth = randomWidth * numRandomSources;
+
+  // Only create the random register once.
+  if (randomValueAndRemain.first == nullptr) {
+    // Declare the random register.
+    auto randReg = regBuilder.create<sv::RegOp>(
+        reg.getLoc(), regBuilder.getIntegerType(randomRegWidth),
+        /*name=*/regBuilder.getStringAttr("_RANDOM"),
+        /*inner_sym=*/
+        regBuilder.getStringAttr(ns.newName(Twine("_RANDOM"))));
+
+    randomValueAndRemain = {randReg, randomRegWidth};
+
+    SmallString<32> randomRegAssign("{{0}} = {");
+
+    // Fill the random register by concatenating calls to `RANDOM in a verbatim
+    // string.
+    for (uint64_t i = 0; i < numRandomSources; ++i) {
+      randomRegAssign.append("`RANDOM");
+      if (i < numRandomSources - 1)
+        randomRegAssign.append(",");
+    }
```

Comment:
This can use `llvm::interleave` to do the same thing. That may arguably be less readable, though. (`llvm::interleaveComma` is really what this is doing, but it's annoying because it needs a stream.)

Reply:
Will stick to a loop, since it makes my line breaking check simpler.